### PR TITLE
Add cache support for Sdk{Runnable|Container}Tasks

### PR DESCRIPTION
**Pull request by narape**
# TL;DR
Add support for caching `SdkRunnableTask` and `SdkContainerTask`

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
The attributes:
* isCached
* getCacheVersion
* isCacheSerializable

Where added to `SdkRunnableTask` and `SdkContainerTask`, The `ProjectClosure` is modified to propagate these properties to TaskTemplate and finally ProtoUtil was also modified to translate from (api) TaskTemplate to proto TaskTemplate. The properties `isCached` and `getCacheVersion` where renamed in TaskTemplate to `discoverable` and `discoveryVersion` to match the one in proto TaskTemplate.

Finally the examples were modified to add caching.

### CODE REVIEW

**Comment by narape** [comment]
Supersedes https://github.com/flyteorg/flytekit-java/pull/121

**Comment by narape** [comment]
@kiarash-rezahanjani I took your PR and expanded. I renamed some things and also added cache to `SdkContainerTasks`

**Review by lupasarin — APPROVED** [review]
